subsystem/drivetrain.py

```diff
@@ -13,17 +25,90 @@
-    m_move: SparkMax
-    m_turn: SparkMax
-    encoder: ctre.CANCoder
-    absolute_encoder_zeroed_pos: float = 0  # Radians
-    drive_reversed: bool = False
-    turn_reversed: bool = False
+
+    def __init__(self, m_move: SparkMax, m_turn: SparkMax, encoder: ctre.CANCoder, encoder_zeroed_absolute_pos: float = 0, drive_reversed: bool = False, turn_reversed: bool = False):
+        self.m_move = m_move
+        self.m_turn = m_turn
+        self.encoder = encoder
+        self.encoder_zeroed_absolute_pos = encoder_zeroed_absolute_pos
+        self.drive_reversed = drive_reversed
+        self.turn_reversed = turn_reversed
 
     def init(self):
         super().init()
-        self.m_turn.init()
         self.m_move.init()
+        self.m_turn.init()
+
+    # make the turn motor set their sensor 0 to current horizontal thingy
+    def zero(self):
+        diff_rad = self.encoder_zeroed_absolute_pos - self.encoder.getAbsolutePosition()
+        
+        # diff_rad = diff * 2 * math.pi
+        current_position_rad = self.get_current_motor_angle()
+        real_position = current_position_rad - diff_rad
+        self.m_turn.set_sensor_position(
+            (real_position / (2 * math.pi) * constants.drivetrain_turn_gear_ratio)
+        )
+
+    # reposition the wheels
+    def set_motor_angle(self, pos: radians):
+        if self.turn_reversed:
+            pos *= -1
+        self.m_turn.set_target_position(
+            (pos / (2 * math.pi)) * constants.drivetrain_turn_gear_ratio
+        )
+
+    def direct_set_motor_angle(self, pos: radians):
+        self.m_turn.set_target_position(
+            (pos / (2 * math.pi)) * constants.drivetrain_turn_gear_ratio
+        )
+
+    def get_current_motor_angle(self) -> radians:
+        return self.m_turn.get_sensor_position() / constants.drivetrain_turn_gear_ratio
+
+    # rotate the wheel so the robot moves
+    def set_motor_velocity(self, vel: meters_per_second):
+        if self.drive_reversed:
+            vel *= -1
+        self.m_move.set_target_velocity(vel * constants.drivetrain_move_gear_ratio)
+
+    def get_motor_velocity(self) -> radians_per_second:
+        return self.m_move.get_sensor_velocity() / constants.drivetrain_move_gear_ratio
+
+class Drivetrain(SwerveDrivetrain):
+    n_00 = SparkMaxSwerveNode(
+        SparkMax(1, config=MOVE_CONFIG),
+        SparkMax(2, config=TURN_CONFIG),
+        wpilib.AnalogEncoder(1),
+        
+    )
+    n_01 = SparkMaxSwerveNode(
+        SparkMax(3, config=MOVE_CONFIG),
+        SparkMax(4, config=TURN_CONFIG),
+        wpilib.AnalogEncoder(0),
+    )
+
+    n_10 = SparkMaxSwerveNode(
+        SparkMax(5, config=MOVE_CONFIG),
+        SparkMax(6, config=TURN_CONFIG),
+        wpilib.AnalogEncoder(2),
+        
+    )
+    n_11 = SparkMaxSwerveNode(
+        SparkMax(7, config=MOVE_CONFIG),
+        SparkMax(8, config=TURN_CONFIG),
+        wpilib.AnalogEncoder(3),
+    )
+
+    # axis_dx = Keymap.Drivetrain.DRIVE_X_AXIS
```

Comment:
Just to clarify, you will eventually need to uncomment all these lines and put actual values in the constants file.